kubernetes-deployment/maintainlib/etcdfix.py

```diff
@@ -0,0 +1,236 @@
+# Copyright (c) Microsoft Corporation
+# All rights reserved.
+#
+# MIT License
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+# to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import yaml
+import os
+import sys
+import subprocess
+import jinja2
+import argparse
+import paramiko
+import kubernetes
+import etcd
+import common
+import logging
+import logging.config
+
+
+
+class etcdfix:
+
+    """
+    A class to reconfiguration etcd. Fix the issue when etcd's data is corrupted
+    """
+
+    def __init__(self, cluster_config, node_config, clean):
```

Comment:
>cluster_config, node_config [](start = 23, length = 27)

can you add ut to express you expectation of the input?
cluster_config['mastermachinelist']
cluster_config['clusterinfo']['etcd_cluster_ips_peer'] =         self.cluster_config['clusterinfo']['etcd-initial-cluster-state']

bad_node_config['etcdid']
bad_node_config['hostip']

Reply:
Why should we add ut to test the correctness of the cluster configuration? This function is designed object initialization. I don't think we should add ut here.

Reply:
sorry, it is more than just ut. you should handle the format incorrectness in your code. Currently, your code implicitly assumes cluster_config has some fields available. e.g., ['mastermachinelist'], what if it doesn't have this field ? what is the expected behavior of your code?  and you should add ut to explicitly test your code's behavior under such ill defined input and meanwhile let code readers understand your purpose in the UT.

---
In reply to: [172112809](https://github.com/Microsoft/pai/pull/345#discussion_r172112809) [](ancestors = 172112809)

Reply:
Let me try to figure out how to add ut. Close this PR currently.

Reply:
Reopen this PR, and developing it in the status of Openning.

Reply:
Add validation class